File: src/contest/413/Q2.ts
export function resultsArray1(queries: number[][], k: number): number[] {
    const arr: number[] = [];

    const res = [];

    for (const [x, y] of queries) {
        const origin = Math.abs(x) + Math.abs(y);

        insertUsingBinarySearch1(arr, origin);

        if (k - 1 < arr.length) {
            res.push(arr[k - 1]);
        } else {
            res.push(-1);
        }
    }

    return res;
}

function insertUsingBinarySearch1(arr: number[], value: number) {
    let low = 0;
    let high = arr.length;

    while (low < high) {
        let mid = Math.floor((low + high) / 2);
        if (arr[mid] < value) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    arr.splice(low, 0, value);
}

export function resultsArray(queries: number[][], k: number): number[] {
    const heap = new MyHeap();
    const res = [];

    for (let i = 0; i < queries.length; i++) {
        const [x, y] = queries[i];
        const math = Math.abs(x) + Math.abs(y);

        if (heap.size() < k - 1) {
            res.push(-1);
            heap.push(math);
        } else if (heap.size() === k - 1) {
            heap.push(math);
            res.push(heap.head());
        } else {
            if (math < heap.head()) {
                heap.changeHead(math);
            }
            res.push(heap.head());
        }
    }

    return res;
}

class MyHeap {
    compare: (a: number, b: number) => boolean;
    ary: number[];
    constructor(compare = (a, b) => a > b) {
        this.compare = compare;
        this.ary = [];
    }

    changeHead(val) {
        this.ary[0] = val;
        this.down(0);
    }

    size() {
        return this.ary.length;
    }

    push(val) {
        this.ary.push(val);
        this.up(this.ary.length - 1);
    }

    head() {
        return this.ary[0];
    }

    up(index) {
        if (index) {
            const parent = (index - 1) >> 1;
            if (this.compare(this.ary[index], this.ary[parent])) {
                this.swag(index, parent);
                this.up(parent);
            }
        }
    }

    down(index) {
        const left = index * 2 + 1;
        if (left < this.ary.length) {
            let change = left;
            const right = left + 1;
            if (
                right < this.ary.length &&
                this.compare(this.ary[right], this.ary[left])
            ) {
                change = right;
            }
            if (this.compare(this.ary[change], this.ary[index])) {
                this.swag(index, change);
                this.down(change);
            }
        }
    }

    swag(a, b) {
        [this.ary[a], this.ary[b]] = [this.ary[b], this.ary[a]];
    }
}

File: src/contest/413/Q2.go
package c_413

func resultsArray(queries [][]int, k int) []int {
	distances := make([]int, k)
	for i := range distances {
		distances[i] = -1
	}
	results := []int{}

	for i := 0; i < len(queries); i++ {
		x, y := queries[i][0], queries[i][1]
		distance := abs(x) + abs(y)

		distances = insertUsingBinarySearch(distances, distance)

		if distances[k-1] == -1 {
			results = append(results, -1)
		} else {
			results = append(results, distances[k-1])
		}
	}

	return results
}

func insertUsingBinarySearch(arr []int, value int) []int {
	low := 0
	high := len(arr)

	for low < high {
		mid := (low + high) / 2
		if arr[mid] < value {
			low = mid + 1
		} else {
			high = mid
		}
	}
	arr = append(arr[:low+1], arr[low:]...)
	arr[low] = value
	arr = arr[:len(arr)-1]

	return arr
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}
